[xpu][test]port test_foreach for Intel GPU

wincent8 opened the pull request:
In this PR, we port `test_foreach` to support Intel GPUs.

Intel GPU support is enabled using the following approaches, with an effort to preserve the original code structure and style as much as possible:

- Use torch.accelerator to provide a generic GPU abstraction.
- Skip test cases when running on XPU if there are known issues.
- Extend CUDA-only test cases to run on both CUDA and XPU.

wincent8 committed:
enable test_foreach for Intel GPU

wincent8 committed:
enable test_foreach for Intel GPU

daisyden reviewed (approved):
good for me.

guangyey commented on test/test_foreach.py:
```diff
@@ -42,11 +43,16 @@
+device_type = (
+    acc.type if (acc := torch.accelerator.current_accelerator(True)) else "cpu"
+)
+TEST_MULTIACCELERATOR = torch.accelerator.device_count() >= 2
```
Please use the one already defined in torch/testing/_internal/common_utils.py

guangyey commented on test/test_foreach.py:
```diff
@@ -1063,6 +1069,8 @@ def test_big_num_tensors(self, device, dtype, op, use_cuda_graph, w_empty):
                     zero_size=False,
                     **kwargs,
                 )
+            elif "xpu" in device:
```
we should check the device is not cuda, then skip CUDAGraph test.

guangyey reviewed (approved):
Some nits. Overall LGTM.

wincent8 committed:
modify according to review

wincent8 committed:
modify according to review

wincent8 commented:
@pytorchbot rebase

wincent8 committed:
update according to review

wincent8 committed:
update according to review

wincent8 commented:
@pytorchbot  rebase

guangyey commented:
Hi @ezyang, may I know if you have any comments on this PR.

wincent8 commented:
@pytorchbot rebase

wincent8 committed:
fix the issue for new added cases

ezyang commented:
@pytorchbot merge

guangyey commented:
Thanks~
@pytorchbot rebase

wincent8 committed:
fix a typo

ezyang committed:
[MERGEDOG] Merge main into PR branch

Resolved conflicts:
  - test/test_foreach.py: Two conflicts, both around test decorators that the
    PR is converting from @onlyCUDA to @onlyOn(["cuda", "xpu"]) to enable
    foreach tests for Intel GPU.
    1. Around test_unary_op_tensors_on_different_devices: kept the PR's
       @onlyOn(["cuda", "xpu"]) decorator and adopted main's typo fix in the
       preceding comment ('cpu]. -> 'cpu'].).
    2. Around test_inplace_foreach_leaf_check_and_grad_fn: kept the PR's
       @onlyOn(["cuda", "xpu"]) decorator and incorporated main's newly added
       test_foreach_max_empty_tensor_error (no overlap with the PR's
       changes).

ezyang commented:
@pytorchbot merge -i

ezyang commented:
## mergedog handoff

All required CI is green. Ready for human review and `@pytorchbot merge`.

During shepherding, claude was invoked **1** time. If any required CI shows red, claude judged that failure unrelated to this PR's changes — please verify below before merging.

### Session 1 — fix-CI (2026-05-03T13:35:51+00:00)

- **Before:** `7e7db321ee41`
- **Verdict:** judged failures spurious (no commit) — failing jobs: libtorch-linux-jammy-cuda13.0-py3.10-gcc11-debug / build, linux-jammy-cuda13.0-py3.10-gcc11-no-ops / build, linux-jammy-cuda13.0-py3.10-gcc11 / build, linux-jammy-rocm-py3.10-mi355 / test (default, 5, 6, linux.rocm.gpu.gfx950.1), linux-noble-xpu-n-py3.10 / test (default, 8, 12, linux.idc.xpu), linux-noble-xpu-n-py3.10 / test (default, 2, 12, linux.idc.xpu), linux-noble-xpu-n-py3.10 / test (default, 10, 12, linux.idc.xpu), linux-noble-xpu-n-py3.10 / test (default, 12, 12, linux.idc.xpu)

<details><summary>claude transcript</summary>

```
claude session 5f980422
claude → Bash: cat /Users/ezyang/.mergedog/contexts/169354.md 2>/dev/null | head -200
claude → Bash: git diff --stat HEAD~1 HEAD 2>/dev/null | head -30
claude → Bash: git diff --stat origin/main...HEAD 2>/dev/null | tail -20
claude: The PR only touches `test/test_foreach.py` and `torch/testing/_internal/common_methods_invocations.py` (porting test_foreach to XPU). None of the failing tests/jobs are in those files:
claude: - The 3 build jobs (libtorch-cuda13-debug, cuda13-no-ops, cuda13) — log tails show only git config noise; Dr. CI flags `cuda13-no-ops` as `"Final attempt failed. Child_process exited with error code 1"` (infra retry exhaustion).
claude: - ROCm `test_triton_heuristics` — Dr. CI marks as FLAKY with similar failure on trunk.
claude: - XPU 2/12, 8/12, 10/12 — Dr. CI marks as BROKEN TRUNK (test_max_autotune, test_autoheuristic, test_triton_heuristics on trunk merge base).
claude: - XPU 12/12 `test_amp_autocast_xpu` — this is in `test/dynamo/test_ctx_manager.py`, not touched by this PR; warning shows XPU autocast doesn't support float64, a pre-existing XPU-specific limitation unrelated to foreach.
claude: No commit — failures are spurious/infra/broken-trunk, not caused by this PR.
claude finished: success (cost $1.4013)
```

</details>

ezyang commented:
@pytorchbot merge -i
